src/antidote_crdt_orset.erl

```diff
@@ -82,115 +76,88 @@
+
 -type member() :: term().
+-type token() :: binary().
+-type tokens() :: [token()].
 
 -spec new() -> orset().
 new() ->
     orddict:new().
 
-%% @doc
-%% without parameter: return all existing elements in the `orset()'
-%% {fragment, elem}: create and return a new `orset()' with all metadata
-%% of an element
-%% {tokens, elem}: returns all uniques tokens of an element in the set
+%% @doc return all existing elements in the `orset()'.
 -spec value(orset()) -> [member()].
-value(ORDict) ->
-    orddict:fetch_keys(ORDict).
-
--spec value({fragment, member()}, orset()) -> orset();
-           ({tokens, member()}, orset()) -> [binary()].
-value({fragment, Elem}, ORSet) ->
-    case value({tokens, Elem}, ORSet) of
-        [] ->
-            orddict:new();
-        Tokens ->
-            orddict:store(Elem, Tokens, orddict:new())
-    end;
-value({tokens, Elem}, ORSet) ->
-    case orddict:find(Elem, ORSet) of
-        error ->
-            [];
-        {ok, Tokens} ->
-            Tokens
-    end;
-value(_, ORSet) ->
-    value(ORSet).
+value(ORSet) ->
+    orddict:fetch_keys(ORSet).
 
 %% @doc generate downstream operations.
-%% If the operation is add or add_all, generate unique tokens for each element
-%% If the operation is remove or remove_all, fetches all unique tokens for
-%% these elements existing in the `orset()'.
--spec downstream(orset_op(), orset()) -> {ok, orset_op()}.
-downstream({add, Elem}, _ORDict) ->
+%% If the operation is add or add_all, generate unique tokens for
+%% each element and fetches the current supporting tokens.
+%% If the operation is remove or remove_all, fetches current
+%% supporting tokens of these elements existing in the `orset()'.
+-spec downstream(orset_op(), orset()) -> {ok, downstream_op()}.
+downstream({add, Elem}, ORSet) ->
     Token = unique(),
-    {ok, {add, {Elem, [Token]}}};
-downstream({add_all, Elems}, _ORDict0) ->
-    DownstreamOp = lists:foldl(fun(Elem, Sum) ->
-                                       Token = unique(),
-                                       Sum ++ [{Elem, [Token]}]
-                               end, [], Elems),
-    {ok, {add_all, DownstreamOp}};
-downstream({remove, Elem}, ORDict) ->
-    ToRemove = value({tokens, Elem}, ORDict),
-    {ok, {remove, {Elem, ToRemove}}};
-downstream({remove_all, Elems}, ORDict) ->
-    ToRemove = lists:foldl(fun(Elem, Sum) ->
-                                   Sum ++ [{Elem, value({tokens, Elem}, ORDict)}]
-                           end, [], Elems),
-    {ok, {remove_all, ToRemove}};
-downstream({reset, {}}, ORDict) ->
+    ToRemove = get_supporting_tokens(Elem, ORSet),
+    DownstreamOp = {Elem, [Token], ToRemove},
+    {ok, [DownstreamOp]};
+downstream({add_all, Elems}, ORSet) ->
+    DownstreamOps = sets:fold(
+        fun(Elem, Ops) ->
+            {ok, [Op]} = downstream({add, Elem}, ORSet),
+            [Op | Ops]
+        end,
+        [],
+        sets:from_list(Elems)
+    ),
```

Comment:
It looks like this fold has quadratic running time, where the old version run in linear time. The optimization to use a set and avoid duplicates is nice though. Maybe you can use `lists:usort` and then calculate all the downstream-ops in one iteration over the OrSet (that's what I did in the remove-wins set). 

Also, it probably is enough to generate one token for all added elements.